runtime/bcutil/cfreader.c

```diff
@@ -2344,27 +2348,50 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 	NEXT_U16(classfile->majorVersion, index);
 	NEXT_U16(classfile->constantPoolCount, index);
 
-	if(classfile->constantPoolCount < 1) {
+	constantPoolAllocationSize = classfile->constantPoolCount;
+
+	if(constantPoolAllocationSize < 1) {
 		errorCode = J9NLS_CFR_ERR_CONSTANT_POOL_EMPTY__ID;
 		offset = index - data - 2;
 		goto _errorFound;
 	}
 
 	VERBOSE_START(ParseClassFileConstantPool);
 	/* Space for the constant pool */
-	if(!ALLOC_ARRAY(classfile->constantPool, classfile->constantPoolCount, J9CfrConstantPoolInfo)) {
+
+	if (J9_ARE_ANY_BITS_SET(findClassFlags, J9_FINDCLASS_FLAG_ANON)) {
+		/* Pre-emptively add new entry to the end of the constantPool for the modified anonClassName.
+		 * If it turns out we dont need it, simply reduce the constantPoolCount by 1, which is
+		 * cheaper than allocating twice.
+		 *
+		 * Can't modify the classfile->constantPoolCount until after readPool is called so
+		 * the classfile is properly parsed. Instead use a temp variable to track the real
+		 * size for allocation, and re-assign to classfile->constantPoolCount later.
+		 *
+		 * If the size of the constantPool is MAX_CONSTANT_POOL_SIZE then we have no choice
+		 * but to reuse the existing classNameUTf cpEntry */
```

Comment:
This is a bug waiting to happen. We should create an issue to find a solution later. Also add an exception tracepoint if this condition occurs.

Reply:
I second that.  It would be better to fail the class load and throw an exception (InternalError?  Something more appropriate?) than to keep this kind of hack.